src/main/java/requestMgmt/RequestManager.java

```diff
@@ -313,6 +320,168 @@ public void fetchPermissionConstants() {
+            int status = response.getStatus();
+            if (status == 200) {
+                this.dataManager.getMetadataManager().fileDeleted(fileID);
+            } else {
+                this.incorrectResponseStatusHandler.handleInvalidResponse(status, "Failed to delete file from server: " + status);
+            }
+        }, requestSendErrorHandler);
+        this.wsManager.sendAuthenticatedRequest(deleteFileReq);
+    }
+
+    /**
+     * Renames the given project on the server.
+     * Also renames the project in its corresponding metadata file.
+     *
+     * @param projectID
+     * @param newName
+     */
+    public void renameProject(long projectID, String newName, String newPath) {
+        Request renameProjectReq = new ProjectRenameRequest(projectID, newName).getRequest(response -> {
+            int status = response.getStatus();
+            if (status == 200) {
+                ProjectMetadata pMeta = this.dataManager.getMetadataManager().getProjectMetadata(projectID);
+                pMeta.setName(newName);
+                this.dataManager.getMetadataManager().putProjectMetadata(newPath, pMeta);
+            } else {
+                this.incorrectResponseStatusHandler.handleInvalidResponse(status, "Failed to rename project to \"" + newName +
+                        "\" on server.");
+            }
+        }, requestSendErrorHandler);
+        this.wsManager.sendAuthenticatedRequest(renameProjectReq);
+    }
+
+    public void sendFileChanges(long fileID, String[] changes, long baseFileVersion) {
+        MetadataManager mm = this.dataManager.getMetadataManager();
+
+        FileMetadata fMeta = mm.getFileMetadata(fileID);
+        ProjectMetadata pMeta = mm.getProjectMetadata(mm.getProjectIDForFileID(fileID));
+        String projRootPath = mm.getProjectLocation(pMeta.getProjectID());
+
+        Request req = getFileChangeRequest(fMeta, changes, response -> {
+            fMeta.setVersion(((FileChangeResponse) response.getData()).getFileVersion());
+            this.dataManager.getMetadataManager().writeProjectMetadataToFile(pMeta, projRootPath,
+                    CoreStringConstants.CONFIG_FILE_NAME);
+        }, null, 1);
+
+        try {
+            this.wsManager.sendRequest(req);
+        } catch (ConnectException e) {
+            System.out.println("Failed to send change request.");
+            e.printStackTrace();
+        }
+
+    }
+
+    private Request getFileChangeRequest(FileMetadata fileMeta, String[] changes, IResponseHandler respHandler,
+                                         IRequestSendErrorHandler sendErrHandler, int retryCount) {
+
+        return new FileChangeRequest(fileMeta.getFileID(), changes, fileMeta.getVersion()).getRequest(response -> {
+
+            // If we failed the first time around, update the fileVersion and
+            // retry.
+            if (response.getStatus() == 409 && retryCount > 0) {
```

Comment:
So it's the same on that other branch but I don't actually think it would loop forever would it? The recursive call decrements the retry count every recursion.